lib/offline/OfflineStorage_Room.cpp

```diff
@@ -684,227 +748,300 @@ namespace MAT_NS_BEGIN
+            ThrowRuntime(env, "call store");
+
+            bool shouldCheck = false;
+            auto current = m_checkAfterInsertCounter.load();
+            size_t newValue;
+            // atomic decrement counter
+            do
             {
-                shouldCheck = false;
-                newValue = current - count;
-            }
-        } while (!m_checkAfterInsertCounter.compare_exchange_weak(current, newValue));
-        if (shouldCheck)
-        {
-            auto sizeEstimate = GetSizeInternal(env);
-            double ratio = static_cast<double>(sizeEstimate) / static_cast<double>(m_size_limit);
-            if (m_notify_fraction <= ratio)
+                if (m_checkAfterInsertCounter <= count)
+                {
+                    shouldCheck = true;
+                    newValue = CHECK_INSERT_COUNT;
+                }
+                else
+                {
+                    shouldCheck = false;
+                    newValue = current - count;
+                }
+            } while (!m_checkAfterInsertCounter.compare_exchange_weak(current, newValue));
+            if (shouldCheck)
             {
-                auto now = PAL::getMonotonicTimeMs();
-                if (now > m_storageFullNotifyTime + m_storageFullNotifyInterval)
+                auto sizeEstimate = GetSizeInternal(env);
+                double ratio =
+                    static_cast<double>(sizeEstimate) /
+                    static_cast<double>(m_size_limit);
+                if (m_notify_fraction <= ratio)
+                {
+                    auto now = PAL::getMonotonicTimeMs();
+                    if (now > m_storageFullNotifyTime + m_storageFullNotifyInterval)
+                    {
+                        m_storageFullNotifyTime = now;
+                        DebugEvent evt;
+                        evt.type = DebugEventType::EVT_STORAGE_FULL;
+                        evt.param1 = std::max<long long int>(0,
+                                                             std::llround(100.0 * ratio));
+                        m_manager.DispatchEvent(evt);
+                    }
+                }
+                if (sizeEstimate >= m_size_limit)
                 {
-                    m_storageFullNotifyTime = now;
-                    DebugEvent evt;
-                    evt.type = DebugEventType::EVT_STORAGE_FULL;
-                    evt.param1 = std::max<long long int>(0, std::llround(100.0 * ratio));
-                    m_manager.DispatchEvent(evt);
+                    ResizeDbInternal(env);
                 }
             }
-            if (sizeEstimate >= m_size_limit)
+            return count;
+        }
+        catch (const std::runtime_error& error)
+        {
+            auto what = error.what();
+            if (!what)
             {
-                ResizeDbInternal(env);
+                what = "*nothing*";
             }
+            LOG_ERROR("Exception in StoreRecords: %s", what);
```

Comment:
Is this sufficient information for debugging the exceptions? We will know `what` the exception is, but we won't know where it came from given the large scope of the try-block.

Reply:
Applies globally in this file.

---
In reply to: [583813385](https://github.com/microsoft/cpp_client_telemetry/pull/790#discussion_r583813385) [](ancestors = 583813385)

Reply:
In automation, we can see device logs, and this should be enough information to tell us where things went horribly wrong. In production, we won't see this at all because the catch block will do what catch blocks are supposed to do, and the exception will be a non-event.